grails-app/controllers/nl/hyve/gb/backend/QueryController.groovy

```diff
@@ -0,0 +1,169 @@
+import nl.hyve.gb.backend.exception.InvalidRequestException
+import nl.hyve.gb.backend.representation.QueryRepresentation
+import nl.hyve.gb.backend.user.AuthContext
+import org.grails.web.converters.exceptions.ConverterException
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.http.HttpStatus
+import org.springframework.web.bind.annotation.PathVariable
+
+class QueryController {
+
+    static responseFormats = ['json']
+
+    @Autowired
+    AuthContext authContext
+
+    @Autowired
+    QueryService queryService
+
+
+    /**
+     * GET /queries
+     *
+     * @returns the list of all queries saved by the user.
+     */
+    def index() {
+        def queries = queryService.list(authContext.user)
+
+        response.status = HttpStatus.OK.value()
+        response.contentType = 'application/json'
+        response.characterEncoding = 'utf-8'
+        BindingHelper.writeValue(response.outputStream, [queries: queries])
+    }
+
+    /**
+     * GET /queries/{id}
+     *
+     * @param id the id of the saved query
+     * @returns the saved query with the provided id, if it exists and is owned by the user,
+     *      status 404 (Not Found) or 403 (Forbidden) otherwise.
+     */
+    def get(@PathVariable('id') Long id) {
+        def query = queryService.get(id, authContext.user)
+
+        response.status = HttpStatus.OK.value()
+        response.contentType = 'application/json'
+        response.characterEncoding = 'utf-8'
+        BindingHelper.writeValue(response.outputStream, query)
+    }
+
+    protected static QueryRepresentation getUserQueryFromString(String src) {
+        if (src == null || src.trim().empty) {
+            throw new InvalidArgumentsException('Empty user query.')
+        }
+        try {
+            try {
+                QueryRepresentation userQuery = BindingHelper.read(src, QueryRepresentation.class)
+                userQuery
+            } catch (JsonProcessingException e) {
+                throw new InvalidArgumentsException("Cannot parse query parameter: ${e.message}", e)
+            }
+        } catch (ConverterException c) {
+            throw new InvalidArgumentsException('Cannot parse query parameter', c)
+        }
+    }
+
+    /**
+     * Deserialises the request body to a user query representation object using Jackson.
+     *
+     * @returns the user query representation object is deserialisation was successful;
+     * responds with code 400 and returns null otherwise.
+     */
+    protected QueryRepresentation bindUserQuery() {
+        if (!request.contentType) {
+            throw new InvalidRequestException('No content type provided')
+        }
+        MimeType mimeType = new MimeType(request.contentType)
+        if (mimeType != MimeType.JSON) {
+            throw new InvalidRequestException("Content type should be ${MimeType.JSON.name}; got ${mimeType}.")
+        }
+
+        try {
+            def src = BindingHelper.write(request.JSON)
```

Comment:
This parses the contents of the body first (`request.JSON`) and then serialises it again to text. You could also use `request.inputStream` as in the [ExportController](https://github.com/thehyve/transmart-core/blob/dev/transmart-rest-api/grails-app/controllers/org/transmartproject/rest/ExportController.groovy#L81) of Transmart.

Reply:
ah true, I just noticed that in Transmart it was like this because of the ApiAuditInterceptor issue, it is changed now